IRSA-6111: Make a compare sources tab in Euclid

jaladh-singhal opened the pull request:
Fixes [IRSA-6111](https://jira.ipac.caltech.edu/browse/IRSA-6111)
Also see: https://github.com/IPAC-SW/irsa-ife/pull/351

- Refactored CombineChart to make it:
   - independent of PinnedChartsContainer, and accept `chartIds` instead
   - handle promises of chartIds (because async fetching of charts is required in Euclid) as well as the loading and error states
   - generic overall and more meaningful names of variables
- Created two specific HOCs of CombineChart:
  1. `CombinePinnedCharts` - to allow existing functionality
  2. `CombineRowCharts` - to allow fetching charts from the selected rows in active table (Euclid's compare sources functionality)
- Bonus: Improved trace titles creation in combined chart - added chart title to trace titles. Old titles that just rely on trace numbering were meaningless because it flattens the traces list nested in charts list so there's no way of telling which trace belong to which chart.

## Testing
### Euclid
https://irsa-6111-compare-sources.irsakudev.ipac.caltech.edu/applications/euclid/

Search by ID -> select Object ID -> Enter `2354019577405107692 2353963395405226961 2353684508405256537 2353952946405260405 2353677340405271646 2354322758405276511 2354215274405166151 2354251826405010397 2353704547404873714 2353721932404932508 2353809153405305199 2353776897404952855 2354281144404971534 2360659649405341842 2360734814404968938 2360651232404669390 2360358793404978811 2360588190404979321 2360776044404861072 2360598910405204293 2360764117405221817 2360527606404894927 2360545771404908021 2360231177404684196 2360503632404903443 2360507038404979260 2360725058404914991 2360357456404777107 2360787580404802683 2360367420404806206 2360734420405222743 2360352268405250703 2360826047404674311 2360513182404968897 2360526211404698268 2360452114404699115 2360882651404702544 2360483160404710523 2360319015404906941 2360582403405020934 2360913088405117008 2360886985404772880 2360311475404993722 2360349765404996732 2360842141405003125 2360238622405013008 2360542602405014725 2360332198405018273 2360625161405264547 2360437631405268102` -> search

#### Main functionality
Select 4-5 rows randomly, and click on venn-diagram button in the table toolbar -> dialog should open with cascading and chart options -> enable cascading -> "ok" should create a combined spectrum in the pinned charts tab. 

#### Edge cases:
- Select all rows in table and click on the compare sources button - it should show an info popup about limit exceeded.
- Add a source ID to the search that doesn't have spectrum associated (I don't know specific example), it should show alert in combine chart dialog that it can't retrieve spectrum (tested this locally)

### Firefly
https://fireflydev.ipac.caltech.edu/irsa-6111-compare-sources/firefly/

Regression test the combine pinned charts functionality: Follow the instructions in https://github.com/Caltech-IPAC/firefly/pull/1499 

Bonus: when switching to single chart view, you can see the trace titles are much clearer than they used to be.

dencoded commented on src/firefly/js/charts/ui/CombineChart.jsx:
```diff
@@ -26,75 +26,136 @@ import {canUnitConv} from '../dataTypes/SpectrumUnitConversion.js';
+        : null;
+};
+
+CombineChart.propTypes = {
```
is it poor-man TypeScript like?

loitly commented on src/firefly/js/charts/ui/CombineChart.jsx:
```diff
@@ -113,62 +174,95 @@ function createTableModel(showAll, tbl_id) {
     const tbl_id = 'combinechart-tbl-id';
     const groupKey = 'combinechart-props';
 
+    const [resolvedChartIds, setResolvedChartIds] = React.useState([]);
+    const [rejectedChartReasons, setRejectedChartReasons] = React.useState([]);
```
`useState` is used throughout our code.  Ok to import it directly.

loitly commented on src/firefly/js/charts/ui/CombineChart.jsx:
```diff
@@ -113,62 +174,95 @@ function createTableModel(showAll, tbl_id) {
-        const selTbl = createTableModel(showAll, tbl_id);
-        dispatchTableAddLocal(selTbl, undefined, false);
-    }, [showAll]);
+        if (chartIds[0] instanceof Promise) {
```
May need to handle cases when `chartIds` is undefined to avoid unexpected error.

jaladh-singhal replied:
I think I can just filter them out of resolvedChartIds, not sure what to show for error because I won't have error message that it's undefined for which row.

Or maybe I should handle handle the undefined check in fetch function to explicitly throw a Error with all the info/

loitly replied:
Ah, I see that CombineChartDialog is not exported.  That's how you can expect `chartIds` to have valid values.  Ignore my comment.

loitly commented on src/firefly/js/charts/ui/CombineChart.jsx:
```diff
@@ -113,62 +174,95 @@ function createTableModel(showAll, tbl_id) {
+    }, [chartIds]);
+
+
+    const chartIdsToCombine = allowChartSelection
+        ? [selectedChartId ?? chartIds[0], ...tblSelectedChartIds] //selectedChartId isn't present in table
+        : resolvedChartIds; //TODO: run chartIds through canCombine() and show warning if not?
```
I understand what `allowChartSelection` is for now.  Maybe it should be called `requiredChartSelection`?  Just a suggestion.

jaladh-singhal replied:
This is actually a flag to control whether to render a table inside dialog that allows further selection among the passed chart IDs. I think I should rename it to `showChartSelectionTable` to resolve ambiguity - what do you think?

loitly reviewed (approved):
Well done!  Good refactoring work.  New functionality nicely integrated into existing code.  Here are a few comments I have while testing.  You and maybe @dencoded should talk about it to decide what to do with them.

To Consider fixing in new tickets:
- Add an option to hide the non-functional 'Show table' button.
- Chart grid (for multiple charts) is clipped at the bottom, hiding the x-axis. Likely a Euclid Results-Viewer issue, not related to this PR.
- Make the 'Table|Spectrum' toggle in the 'Spectra' tab a generic Chart function to switch between data table and chart views.

Should Fix:
- Highlighting or deselecting rows refreshes the 'Pinned Chart' unnecessarily.
- Issue occurs when:
  - Loading multiple sources.
  - Comparing selected rows.
  - Toggling highlighted rows or selecting a different row causes 'Pinned Chart' to refresh.

jaladh-singhal commented:
@loitly I've addressed your comments as follows:

> Add an option to hide the non-functional 'Show table' button.

Fixed - pinned charts viewer now shows "show table" button only when traces in the chart have same table, which is never the case in combined chart.

> Chart grid (for multiple charts) is clipped at the bottom, hiding the x-axis. Likely a Euclid Results-Viewer issue, not related to this PR.

@kpuriIpac is fixing it in his PR

> Make the 'Table|Spectrum' toggle in the 'Spectra' tab a generic Chart function to switch between data table and chart views.

Needs a ticket with details on what you intend

> Highlighting or deselecting rows refreshes the 'Pinned Chart' unnecessarily.

Fixed - it was happening because row chart fetching was happening on every (de)highlight and (de)selection. Now I made it to happen only in an effect and if it's not available in a local cache.
